File: testing/centralprocessingunit.cpp
#include "centralprocessingunit.h"
#include<iostream>
#include<cmath>


    double CentralProcessingUnit::Score () const {
        unsigned int sum, var;
        double mem=getMem();
        sum= (num_physical_cores*getClock()+mem);
        if (hypertrheading){
            var= (sum*performance_percentage_increase)/100; //aumento performance percentuale sistemata in modo tale che si traduca in aumento medio del 10% nelle performance del sistema finale
            sum+=var;
        }
        if(mem>4 && mem<=8)
                    var=(sum*20)/100;
                else{
                    if(mem<=4)
                        var=(sum*10)/100;
                    if(mem>8)
                        var=(sum*30)/100;
                }
        sum+=var;
        return sum;
    }

    double CentralProcessingUnit::getScore() const {
        return Score();
    }

    void CentralProcessingUnit::modifyCores (int c) {
        if(c<0 && std::abs(c)>num_physical_cores) std::cerr<<" modify Error, valore comporta numero cores negativo"<<std::endl;
        else num_physical_cores+=c;
    }

    void CentralProcessingUnit::setCores (int c) {
        if(c>0)
        num_physical_cores=c;
        else std::cerr<<" set Error, inserire valore>0 "<<std::endl;
    }

    void CentralProcessingUnit::setHyperT (bool b) {
        hypertrheading=b;
    }

    CentralProcessingUnit::CentralProcessingUnit(int c, int mem, int num, bool b) : BaseComponent(c,mem), num_physical_cores(num), hypertrheading(b) {}


unsigned int CentralProcessingUnit::performance_percentage_increase=10;


std::ostream& operator<< (std::ostream& os , const CentralProcessingUnit& cpu ) {
    double mem=cpu.getMem();
        if(cpu.hypertrheading){
            return os << "La CPU con hyperthreading ha: "<< cpu.num_physical_cores <<" cores "<<cpu.getClock()<< " MHz "<< mem << " MB di cache";
        }
        else
            return os << "La CPU ha: "<< cpu.num_physical_cores<<" cores "<<cpu.getClock()<< " MHz "<< mem << " MB di cache";

}

File: completed/Progetto_pao1718/Java/esecuzione/CentralProcessingUnit.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
//package esecuzione;

/**
 *
 * @author mhetac
 */
public class CentralProcessingUnit extends BaseComponent {
    private static int performance_percentage_increase=10;  //serve perche' con hyperthreading si ha un aumento performace di circa il 10%
    private int num_physical_cores;
    private boolean hypertrheading;
    private double Score () {
        int sum, var=0;
        double mem=getMem();
        sum= (int)((num_physical_cores*getClock())+mem);
        if (hypertrheading){
            var= (sum*performance_percentage_increase)/100; //aumento performance percentuale sistemata in modo tale che si traduca in aumento medio del 10% nelle performance del sistema finale
            sum+=var;
        }
        if(mem>4 && mem<=8)
            var=(sum*20)/100;
        else{
            if(mem<=4)
            	var=(sum*10)/100;
            if(mem>8)
            	var=(sum*30)/100;
        }
        sum+=var;
        return sum;}
    
    public double getScore(){ //override
        return Score();
    }
    public void modifyCores (int c){
        if(num_physical_cores+c<0) System.err.println(" modify Error, valore comporta numero cores negativo");
        else num_physical_cores+=c;
   }
    public void setCores (int c){
        if(c>0)
        num_physical_cores=c;
        else System.err.println(" set Error, inserire valore>0 ");
    }
    public void setHyperT (boolean hy){
        hypertrheading=hy;
        }
   
    public void Print(){  //override
    	double mem=getMem();
            if(hypertrheading){
            	System.out.println("La CPU con hypertreading ha "+num_physical_cores+" cores "+getClock()+" MHz "+mem+" MB di cache");
            	}
            else
            	System.out.println("La CPU ha "+num_physical_cores+" cores "+getClock()+" MHz "+mem+" MB di cache");
        
    }
    
    public CentralProcessingUnit (int clock, double mem, int cores, boolean hy){
        super(clock, mem); //richiama il costruttore della classe padre BaseComponent
        num_physical_cores=cores;
        hypertrheading=hy;
    }
    
    public CentralProcessingUnit(){  //costruttore senza parametri, pone tutto a zero
        //super();  implicita
        num_physical_cores=0;
        hypertrheading=false;
    }
}
